### PR TITLE
Fix parsing of @_unavailableFromAsync

### PR DESCRIPTION
**Pull request by beccadax**
There’s a very easy to reach `llvm_unreachable()` in this code which ought to be a diagnostic, as well as a couple of other issues. Rework it into something that’s a bit better at handling the edge cases.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please test

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -3780,35 +3780,59 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
...
-      consumeToken();
-      if (!consumeIf(tok::colon)) {
-        if (!Tok.is(tok::equal)) {
-          diagnose(Tok.getLoc(), diag::attr_expected_colon_after_label, flag);
-          return makeParserSuccess();
+      Identifier label;
+      SourceLoc labelLoc;
+      parseOptionalArgumentLabel(label, labelLoc, /*isAttr=*/true);
+
+      if (label.empty()) {
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp:3792]: Considering `(_: "message"`, using `labelLoc` is safer.
```suggestion
      if (labelLoc.isInvalid()) {
```
Same for the `newCode` check below

Comment (beccadax) [reply lib/Parse/ParseDecl.cpp:3792]: Added more test cases covering these combos, reworked the code to behave more nicely.

**Comment by beccadax** [comment]
@swift-ci please smoke test

**Comment by beccadax** [comment]
@swift-ci please test
